src/Inventory/Asset/Unmanaged.php

```diff
@@ -0,0 +1,390 @@
+                    $val->mac,
+                    $entities_id
+                );
+                //manage converted object
+                if (!empty($result)) {
+                    $converted_object = new $result['itemtype']();
+                    if ($converted_object->getFromDB($result['id'])) {
+                        $this->item = $converted_object;
+                        $items_id = $result['id'];
+                        $itemtype = $result['itemtype'];
+                    } else {
+                        $need_to_add = true;
+                    }
+                } else {
+                    $need_to_add = true;
+                }
+            } else {
+                $need_to_add = true;
+            }
+
+            if ($need_to_add) {
+                //else add it
+                $input = $this->handleInput($val, $this->item);
+                unset($input['ap_port']);
+                unset($input['firmware']);
+                $items_id = $this->item->add(Sanitizer::sanitize($input));
+                $this->setNew();
+            }
+        }
+
+        if (in_array($itemtype, $CFG_GLPI['agent_types'])) {
+            $this->agent->update(['id' => $this->agent->fields['id'], 'items_id' => $items_id, 'entities_id' => $entities_id]);
+        } else {
+            $this->agent->fields['items_id'] = $items_id;
+            $this->agent->fields['entities_id'] = $entities_id;
+        }
+
+        //check for any old agent to remove
+        $agent = new \Agent();
+        $agent->deleteByCriteria([
+            'itemtype' => $this->item->getType(),
+            'items_id' => $items_id,
+            'NOT' => [
+                'id' => $this->agent->fields['id']
+            ]
+        ]);
+
+        $val->id = $this->item->fields['id'];
+
+        if ($entities_id == -1) {
+            $entities_id = $this->item->fields['entities_id'];
+        }
+        $val->entities_id = $entities_id;
+
+        if ($entities_id != $this->item->fields['entities_id']) {
+            //asset entity has changed in rules; do transfer
+            $doTransfer = \Entity::getUsedConfig('transfers_strategy', $this->item->fields['entities_id'], 'transfers_id', 0);
+            $transfer = new Transfer();
+            if ($doTransfer > 0 && $transfer->getFromDB($doTransfer)) {
+                $item_to_transfer = [$this->itemtype => [$items_id => $items_id]];
+                $transfer->moveItems($item_to_transfer, $entities_id, $transfer->fields);
+                //and set new entity in session
+                $_SESSION['glpiactiveentities']        = [$entities_id];
+                $_SESSION['glpiactiveentities_string'] = $entities_id;
+                $_SESSION['glpiactive_entity']         = $entities_id;
+            } else {
+                //no transfert so revert to old entities_id
+                $val->entities_id = $this->item->fields['entities_id'];
+            }
+        }
+
+        $this->handlePorts();
```

Comment:
Unless I miss something, this method is not defined in `Unmanaged`.

Reply:
Is needed, and defined as ```Trait``` from ```MainAsset```

```php

abstract class MainAsset extends InventoryAsset
{
    use InventoryNetworkPort;

    /** @var array */
    protected $extra_data = [
```

Reply:
Indeed.